fasthttpadaptor/adaptor.go

```diff
@@ -48,95 +51,125 @@ func NewFastHTTPHandlerFunc(h http.HandlerFunc) fasthttp.RequestHandler {
 // according to https://github.com/valyala/fasthttp#switching-from-nethttp-to-fasthttp .
 func NewFastHTTPHandler(h http.Handler) fasthttp.RequestHandler {
 	return func(ctx *fasthttp.RequestCtx) {
+		// transform request
+		// http.Request cannot be reused because it has unexported fields
 		var r http.Request
-
-		body := ctx.PostBody()
 		r.Method = string(ctx.Method())
-		r.Proto = "HTTP/1.1"
-		r.ProtoMajor = 1
-		r.ProtoMinor = 1
-		r.RequestURI = string(ctx.RequestURI())
-		r.ContentLength = int64(len(body))
-		r.Host = string(ctx.Host())
-		r.RemoteAddr = ctx.RemoteAddr().String()
-
-		hdr := make(http.Header)
+		rURL, err := url.ParseRequestURI(string(ctx.RequestURI()))
+		if err != nil {
+			ctx.Logger().Printf("cannot parse requestURI %q: %s", r.RequestURI, err)
+			ctx.Error("Internal Server Error", fasthttp.StatusInternalServerError)
+			return
+		}
+		r.URL = rURL
+		if ctx.Request.Header.IsHTTP11() {
+			r.Proto = "HTTP/1.1"
+			r.ProtoMajor = 1
+			r.ProtoMinor = 1
+		} else {
+			r.Proto = "HTTP/1.0"
+			r.ProtoMajor = 1
+			r.ProtoMinor = 0
+		}
+		r.Header = make(http.Header)
 		ctx.Request.Header.VisitAll(func(k, v []byte) {
 			sk := string(k)
 			sv := string(v)
 			switch sk {
 			case "Transfer-Encoding":
 				r.TransferEncoding = append(r.TransferEncoding, sv)
 			default:
-				hdr.Set(sk, sv)
+				r.Header.Set(sk, sv)
 			}
 		})
-		r.Header = hdr
-		r.Body = &netHTTPBody{body}
-		rURL, err := url.ParseRequestURI(r.RequestURI)
-		if err != nil {
-			ctx.Logger().Printf("cannot parse requestURI %q: %s", r.RequestURI, err)
-			ctx.Error("Internal Server Error", fasthttp.StatusInternalServerError)
-			return
+		reqBody := ctx.PostBody()
+		if reqBody != nil {
+			r.Body = ioutil.NopCloser(bytes.NewReader(reqBody))
+			r.ContentLength = int64(len(reqBody))
 		}
-		r.URL = rURL
+		r.Host = string(ctx.Host())
+		r.RemoteAddr = ctx.RemoteAddr().String()
+		r.RequestURI = string(ctx.RequestURI())
```

Comment:
You're doing `string(ctx.RequestURI())` above as well, might as well combine them for one less allocation.

Reply:
Hello.
I added a stringsBuilder to combine multiple bytes into one string.
I also added BenchmarkAdaptor.

Reply:
Interesting concept the `stringsBuilder` but I'm afraid I don't see the use here? You're just putting everything into one big buffer. That's not what I meant when I said `combine them for one less allocation`. I meant doing `url.ParseRequestURI(r.RequestURI)` like you already changed. How does `stringsBuilder` improve anything?

Reply:
I use the stringsBuilder  to  in the initHTTPRequest function convert the method, uri, host, header to a string. It only needs to allocate memory once.
I think stringbuilder is very important if the header is large.